core/trino-main/src/main/java/io/trino/operator/table/Sequence.java

```diff
@@ -0,0 +1,301 @@
+                            ScalarArgumentSpecification.builder()
+                                    .name(STEP_ARGUMENT_NAME)
+                                    .type(BIGINT)
+                                    .defaultValue(1L)
+                                    .build()),
+                    new DescribedTable(descriptor(ImmutableList.of("sequential_number"), ImmutableList.of(BIGINT))));
+        }
+
+        @Override
+        public TableFunctionAnalysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments)
+        {
+            Object startValue = ((ScalarArgument) arguments.get(START_ARGUMENT_NAME)).getValue();
+            if (startValue == null) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Start is null");
+            }
+
+            Object stopValue = ((ScalarArgument) arguments.get(STOP_ARGUMENT_NAME)).getValue();
+            if (stopValue == null) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Stop is null");
+            }
+
+            Object stepValue = ((ScalarArgument) arguments.get(STEP_ARGUMENT_NAME)).getValue();
+            if (stepValue == null) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Step is null");
+            }
+
+            long start = (long) startValue;
+            long stop = (long) stopValue;
+            long step = (long) stepValue;
+
+            if (start < stop && step <= 0) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, format("Step must be positive for sequence [%s, %s]", start, stop));
+            }
+
+            if (start > stop && step >= 0) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, format("Step must be negative for sequence [%s, %s]", start, stop));
+            }
+
+            return TableFunctionAnalysis.builder()
+                    .handle(new SequenceFunctionHandle(start, stop, start == stop ? 0 : step))
+                    .build();
+        }
+    }
+
+    public record SequenceFunctionHandle(long start, long stop, long step)
+            implements ConnectorTableFunctionHandle
+    {}
+
+    public static ConnectorSplitSource getSequenceFunctionSplitSource(SequenceFunctionHandle handle)
+    {
+        // using BigInteger to avoid long overflow since it's not in the main data processing loop
+        BigInteger start = BigInteger.valueOf(handle.start());
+        BigInteger stop = BigInteger.valueOf(handle.stop());
+        BigInteger step = BigInteger.valueOf(handle.step());
+
+        if (step.equals(BigInteger.ZERO)) {
+            checkArgument(start.equals(stop), "start is not equal to stop for step = 0");
+            return new FixedSplitSource(ImmutableList.of(new SequenceFunctionSplit(start.longValueExact(), stop.longValueExact())));
+        }
+
+        BigInteger splitJump = BigInteger.valueOf(DEFAULT_SPLIT_SIZE - 1).multiply(step);
```

Comment:
One artifact of this logic is that all but one split will be unbalanced. For instance, if there are 1_000_001 elements in the sequence, one will produce 1_000_000 and the other one will produce 1. It may not matter in general, but it could for some specific workloads.

One possible solution to that would be to calculate the number of splits based on a max number of elements per split, and then divide the elements in the sequence equally among that number of splits. So, in the example above, we'd need 2 splits, and the number of elements in each split would be 500_001 and 500_000, respectively.

Reply:
Here's how I'd do it (you'll need to convert it to use BigInteger):

```java
long start = handle.start();
long stop = handle.stop();
long step = handle.step();

stop = (stop - start) / step * step + start; // align to step

long range = stop - start + 1;
long splitCount = range / (maxSplitSize * step) + 1;
long targetRange = (range / splitCount);

targetRange = (targetRange + step) / step * step; // align to ceil(targetRange / step) * step

for (int i = 0; i < splitCount; i++) {
    long splitStart = start + i * targetRange;
    long splitStop = start + (i + 1) * targetRange - step;
    splitStop = step > 0 ? Math.min(stop, splitStop) : Math.max(stop, splitStop);

    ... emit SequenceFunctionSplit(splitStart, splitStop)
}
```

Reply:
I managed to simplify it more with operating on steps instead of ranges.